components/organisation-unit-tree/src/organisation-unit-tree/find-common-orgunit-roots.js

```diff
@@ -0,0 +1,24 @@
+/**
+ * Find the minimum/common root units from a list of orgunits.
+ * This is used to "deduplicate" a list of units, where some unit in the list
+ * may be a parent of another, and thus only the parent should be included as a root.
+ *
+ * This is mainly because of the /me API returning the verbatim selected units,
+ * where the user is able to select children deep in a tree, even if an ancestor above is selected
+ * @param {Array} unitsWithLevel - List of orgunits with their level
+ * @returns {Array} - A filtered list of the minimum root units
+ */
+export const findCommonOrgUnitRoots = (unitsWithLevel) => {
+    const sorted = unitsWithLevel.sort((a, b) => a.level - b.level)
+
+    const minimumRoots = sorted.filter((ou, index, array) => {
+        // since the array is sorted by level we can just check the previous units,
+        // because we want to get the "minimum" level
+        const previousUnits = array.slice(0, index)
+        // if a previous unit has a path that is a prefix of the current path,
+        // then the current path is a child and should not be included
+        return !previousUnits.some((pu) => ou.path.startsWith(pu.path))
```

Comment:
Not really, this is run on the `rootUnits` passed to the component. This is usually just a couple of units - eg. the assigned organisation units to an user from `/me` api. If you're posting thousands of units, something is not right with the configuration of the instance.

Also the optimization of just checking previousUnits would prevent it from being unnecessary expensive. The inner `.some()` would most likely return pretty early due to an identified root. However it's probably possible to optimize it further by adding identified roots.

But again; if you posted thousands of units, you would send a request for each unit - which would cause more trouble than this loop. Please see my sidenote in description about that.